system/ioctl.go

```diff
@@ -136,25 +135,20 @@ func GetMountInfoFromDeviceID(devID [2]uint32) (*MountInfo, error) {
 	return nil, ErrDevNotMounted
 }
 
-// GetBlockDevicePathFromID returns the expanded path to the device with the
+// GetBlockDeviceFromID returns the expanded path to the device with the
 // given device ID, devID, on the form [2]uint32{major, minor}
-func GetBlockDevicePathFromID(devID [2]uint32) (string, error) {
-	path := fmt.Sprintf("/dev/block/%d:%d", devID[0], devID[1])
-
-	path, err := filepath.EvalSymlinks(path)
-	if err != nil {
-		return "", err
-	}
-	return filepath.Clean(path), nil
+func GetBlockDeviceFromID(devID [2]uint32) (string, error) {
+	return sys.DeviceFromID(devID)
 }
 
 // GetDeviceIDFromPath retrieves the device id for the block device pointed to by
 // the inode at path.
 func GetDeviceIDFromPath(path string) ([2]uint32, error) {
-	var stat unix.Stat_t
+	var stat stat
 
-	if err := unix.Stat(path, &stat); err != nil {
-		return [2]uint32{^uint32(0), ^uint32(0)}, err
+	if err := sys.Stat(path, &stat); err != nil {
+		return [2]uint32{^uint32(0), ^uint32(0)}, errors.Wrapf(err,
+			"stat %s", path)
```

Comment:
The code I want to comment on is a little bit further down, but it is from an earlier commit, so I can't put the comment there. However, following our discussion from yesterday, I think we may want to change this code a bit:

```go
	switch devType {
	// If path refers to a special file (e.g. device file under /dev), then
	// st_dev refers to the device number of the mounted devfs. The device
	// number for a special file is under the st_rdev property, ref stat(2).
	case unix.S_IFBLK,
		unix.S_IFCHR,
		unix.S_IFIFO,
		unix.S_IFSOCK:
		return [2]uint32{
			unix.Major(stat.Rdev), unix.Minor(stat.Rdev)}, nil
	// If path refers to a regular file, then st_dev gives the device number
	// of the underlying block device.
	case unix.S_IFDIR,
		unix.S_IFREG,
		unix.S_IFLNK:
		return [2]uint32{
			unix.Major(stat.Dev), unix.Minor(stat.Dev)}, nil
	}
```

1. I think the first case should only cover `S_IFBLK`. No other device is appropriate for taking the `Rdev`. I mean, you *can* do it for a `S_IFCHR` device as well, but it's not valid for snapshots.
2. I would opt for being somewhat restrictive and make the second case only cover `S_IFDIR`, with a `default` that covers the rest with an error message (`S_IFLNK` is implicitly covered by `S_IFDIR`, because you are using `Stat`, not `Lstat`). I think this is less surprising. For example, consider the case where a user tries to do this:
   ```bash
   mender snapshot dump /home/user/my-test-file.fs
   ```
   This will dump the filesystem which the file is on, but it looks more like it should take a snapshot of that file. Instead I think an error message here is fine, since dumping a file is kind of useless (you can just copy it).

So in short, only block devices and directories would be valid arguments to `--source`. What do you think?

Reply:
Yes and no. I agree that the source-flag can be confusing in the file-case. However, this function is in the system package, and therefore should not be tied to requirements in the cli package. If there's a strong sentiment for discarding file and non-device special files from the source flag, then I think the cli package should do the stat and mode-check.
On a different note, for systems running overlay passing a directory to `--source` will wail, but a file will result in the device storing the file being resolved.

Reply:
> However, this function is in the system package, and therefore should not be tied to requirements in the cli package.

Yes, but we put it there, we should not let that rule functionality. After all it's the user experience that counts. I agree that it might fit better in the cli package for this use case, so we can just move it.

> On a different note, for systems running overlay passing a directory to --source will wail, but a file will result in the device storing the file being resolved.

Yes, but this is very unintuitive, and I think I would prefer an error message in this case. If you try to reference a file like this, you don't even know which device you will snapshot. Will you get the the base branch? Or maybe the COW branch, which is nearly empty? It can even change from one invocation to the next if you write to the file in between. I think requiring directories or block devices is cleaner and less confusing.